### PR TITLE
Document and document type read API

### PR DESCRIPTION
**Pull request by kjac**
### Prerequisites

- [x] I have added steps to test this contribution in the description below

### Description

This PR adds read operations for document and document types to the management API.

A few usages of the obsolete `ILocalizationService` have also been replaced with `ILanguageService` because I stumbled upon them while doing the PR 🤓 

### Limitations

Document scheduled publishing and document type list view configuration are conscioulsy omitted from this PR because their data contracts are yet to be defined.

### Breaking changes

We need to break `ContentTypeSort` to add document type keys for the management API output related to allowed child document types.

Eventually we will break it once more when we remove the integer ID, but this cannot be done until we get rid of the content type editor in the current backoffice.

### Special quirkyness

The core currently uses `Json.NET` for serialization, but the management API uses `System.Text.Json`. This causes some issues when reading document properties based on JSON, because the two serializers clash horribly when attempting to serialize the document output from the API.

Eventually we will replace `Json.NET` entirely for V13, but since we are still very much dependent on the current backoffice to create test data, we can't make that switch just yet as it would break the current backoffice. 

To maintain backwards compatability with the current backoffice while still being able to utilize `System.Text.Json` for serializing core data (i.e. document properties), this PR introduces the `ContextualJsonSerializer`. This serializer implementation attempts to be clever about the current request state, thus using `System.Text.Json` for serializing when the current request is within the management API and `Json.NET` otherwise. 

This is purely a temporary measure. It might not work flawlessly - there may be cases where the core ends up caching `System.Text.Json` data in memory and thus breaking later on when attempting to use the cached data for `Json.NET` serialization. We'll have to cross that bridge when we get there. Hopefully it won't be a huge issue moving forwards.

### Testing this PR

1. Verify that documents can be read from `https://localhost:44331/umbraco/management/api/v1.0/document/{key}`
2. Verify that document types can be read from `https://localhost:44331/umbraco/management/api/v1.0/document-type/{key}`

### CODE REVIEW

File: src/Umbraco.Cms.Api.Management/Controllers/Document/ByKeyDocumentController.cs [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    {
+        _contentService = contentService;
+        _documentViewModelFactory = documentViewModelFactory;
+    }
+
+    [HttpGet("{key:guid}")]
+    [MapToApiVersion("1.0")]
+    [ProducesResponseType(typeof(DocumentViewModel), StatusCodes.Status200OK)]
+    [ProducesResponseType(StatusCodes.Status404NotFound)]
+    public async Task<ActionResult<DocumentViewModel>> ByKey(Guid key)
```
Comment (bergmania) [inline src/Umbraco.Cms.Api.Management/Controllers/Document/ByKeyDocumentController.cs:25]: We discussed this yesterday
```suggestion
    public async Task<IActionResult> ByKey(Guid key)
```

File: src/Umbraco.Cms.Api.Management/Controllers/Document/ByKeyDocumentController.cs [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    {
+        IContent? content = _contentService.GetById(key);
```
Comment (bergmania) [inline src/Umbraco.Cms.Api.Management/Controllers/Document/ByKeyDocumentController.cs:27]: Wonder if it is better to hide this behind a private method returning attempt. Thereby we can start the pattern and use the same method in the base for Id not found, as we will need for Update operations.

File: src/Umbraco.Cms.Api.Management/Controllers/DocumentType/ByKeyDocumentTypeController.cs [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    [ProducesResponseType(StatusCodes.Status404NotFound)]
+    public async Task<ActionResult<DocumentTypeViewModel>> ByKey(Guid key)
+    {
+        IContentType? contentType = _contentTypeService.Get(key);
+        if (contentType == null)
+        {
+            return NotFound();
+        }
+
+        DocumentTypeViewModel model = _umbracoMapper.Map<DocumentTypeViewModel>(contentType)!;
```
Comment (bergmania) [inline src/Umbraco.Cms.Api.Management/Controllers/DocumentType/ByKeyDocumentTypeController.cs:33]: Same comments as for documents

File: src/Umbraco.Cms.Api.Management/Controllers/DocumentType/ByKeyDocumentTypeController.cs [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    {
+        _contentTypeService = contentTypeService;
+        _umbracoMapper = umbracoMapper;
+    }
+
+    [HttpGet("{key:guid}")]
+    [MapToApiVersion("1.0")]
+    [ProducesResponseType(typeof(DocumentTypeViewModel), StatusCodes.Status200OK)]
+    [ProducesResponseType(StatusCodes.Status404NotFound)]
+    public async Task<ActionResult<DocumentTypeViewModel>> ByKey(Guid key)
```
Comment (bergmania) [inline src/Umbraco.Cms.Api.Management/Controllers/DocumentType/ByKeyDocumentTypeController.cs:25]:
```suggestion
    public async Task<IActionResult> ByKey(Guid key)
```

File: src/Umbraco.Cms.Api.Management/Factories/ContentUrlFactory.cs [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+        _publishedUrlProvider = publishedUrlProvider;
+    }
+
+    public async Task<IEnumerable<ContentUrlInfo>> GetUrlsAsync(IContent content)
+    {
+        if (_umbracoContextAccessor.TryGetUmbracoContext(out IUmbracoContext? umbracoContext) == false)
+        {
+            _loggerFactory.CreateLogger<ContentUrlFactory>().LogWarning($"Unable to create an Umbraco context while attempting to get URLs for content: {content.Key}");
+            return Array.Empty<ContentUrlInfo>();
+        }
```
Comment (bergmania) [inline src/Umbraco.Cms.Api.Management/Factories/ContentUrlFactory.cs:52]: I think it is okay to explode here, if the httpcontext is not there?
```suggestion
        IUmbracoContext umbracoContext = _umbracoContextAccessor.GetRequiredUmbracoContext();
```

File: src/Umbraco.Cms.Api.Management/ViewModels/ContentType/PropertyTypeContainerViewModelBase.cs [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+public abstract class PropertyTypeContainerViewModelBase
+{
+    public Guid Key { get; set; }
+
+    public Guid? ParentKey { get; set; }
+
+    public string? Name { get; set; }
+
+    // NOTE: This needs to be a string because it can be anything in the future (= not necessarily limited to "tab" or "group")
+    public string Type { get; set; } = string.Empty;
```
Comment (bergmania) [inline src/Umbraco.Cms.Api.Management/ViewModels/ContentType/PropertyTypeContainerViewModelBase.cs:12]: But is it a breaking change to add a value to an enum in the future?

**Review by bergmania — COMMENTED** [review]
Overall it looks good and tests good.. Just added a few comments. Can most likely be fixed up in the Create/Update/Delete PRs

File: src/Umbraco.Cms.Api.Management/ViewModels/ContentType/PropertyTypeContainerViewModelBase.cs (hunk 6 above)
Comment (kjac) [reply src/Umbraco.Cms.Api.Management/ViewModels/ContentType/PropertyTypeContainerViewModelBase.cs:12]: It is not, but the point is we don't want to know about them in the future; they are purely data needed for presentation, we just need to store the data. So sometime down the line we'll get rid of the enum entirely 😄

File: src/Umbraco.Cms.Api.Management/Controllers/Document/ByKeyDocumentController.cs (hunk 2 above)
Comment (kjac) [reply src/Umbraco.Cms.Api.Management/Controllers/Document/ByKeyDocumentController.cs:27]: We haven't used the attempt pattern for simply retrieving items anywhere else; but we do have async Get-methods in the new service implementations. We should do something similiar.

The create/update PR will include a new service for content editing, which is using async + attempt. I'll add a FIXME here and include something in the create/update service for retrieving content items async.

File: src/Umbraco.Cms.Api.Management/Controllers/DocumentType/ByKeyDocumentTypeController.cs (hunk 3 above)
Comment (kjac) [reply src/Umbraco.Cms.Api.Management/Controllers/DocumentType/ByKeyDocumentTypeController.cs:33]: The content type service is likely up for a rewrite for async + attempt. This rewrite should then include a GetAsync, which can be utilized here as well.

I'll add a FIXME for now and a note in the backlog for the content type controller 👍 we'll circle back to it when we get to content types create/update.

**Review by bergmania — APPROVED** [review]

**Pull request merged**